usr/lib/cca_stdll/cca_specific.c

```diff
@@ -4366,6 +4366,204 @@ static CK_RV cca_get_min_card_level(STDLL_TokData_t *tokdata)
     return ret;
 }
 
+static CK_BBOOL cca_get_acp(struct cca_role_data *role_data,
+                            CK_ULONG role_data_size, CK_ULONG acp_num)
+{
+    struct cca_acp_segment *acp_segment;
+    CK_BBOOL ret = CK_FALSE, found = CK_FALSE;
+    CK_BYTE *bitmap;
+    CK_ULONG i, ofs, bit_no;
+
+    if (role_data == NULL || role_data_size < sizeof(struct cca_role_data))
+        goto out;
+
+    ofs = sizeof(struct cca_role_data);
+    for (i = 0; i < role_data->num_segments && ofs < role_data_size; i++) {
+        if (ofs + sizeof(struct cca_acp_segment) > role_data_size)
+            goto out;
+
+        acp_segment =
+              (struct cca_acp_segment *)((CK_BYTE*)role_data + ofs);
+        ofs += sizeof(struct cca_acp_segment);
+
+        if (acp_num >= acp_segment->start_bit_no &&
+            acp_num <= acp_segment->end_bit_no) {
+            if (ofs + acp_segment->num_bytes > role_data_size)
+                goto out;
+
+            bitmap = ((CK_BYTE *)acp_segment) + sizeof(struct cca_acp_segment);
+            bit_no = acp_num - acp_segment->start_bit_no;
+
+            if (ACP_BYTE_NO(bit_no) > acp_segment->num_bytes)
+                goto out;
+
+            ret = (bitmap[ACP_BYTE_NO(bit_no)] & ACP_BIT_MASK(bit_no)) != 0;
+            found = CK_TRUE;
+            goto out;
+        }
+
+        ofs += acp_segment->num_bytes;
+    }
+
+out:
+    TRACE_DEVEL("ACP 0x%04lx: %s %s\n", acp_num, ret ? "enabled" : "disabled",
+                found ? "" : "(not in any segment)");
+
+    return ret;
+}
+
+typedef struct {
+    CK_BBOOL acps_set;
+    struct cca_acp_info acp_info;
+} cca_acp_info_data_t;
+
+/*
+ * Callback function used by cca_get_acp_infos() to determine the
+ * minimum ACP settings among all available APQNs.
+ */
+static CK_RV cca_get_acp_info_handler(STDLL_TokData_t *tokdata,
+                                      const char *adapter,
+                                      unsigned short card,
+                                      unsigned short domain,
+                                      void *handler_data)
+{
+    cca_acp_info_data_t *data = (cca_acp_info_data_t *)handler_data;
+    unsigned char rule_array[CCA_RULE_ARRAY_SIZE] = { 0, };
+    unsigned char role_name[8] = { 0 };
+    unsigned char roles_data[4000], role_data[4000];
```

Comment:
intentionally no init to zeros for roles_data and role_data?

Reply:
yes, those will be set by the CCA verb. No need to initialize them.